Classifier: record tutorials as completed when you click Get Started

eatyourgreens opened the pull request:
- remove `tutorials.tutorialSeenTime` and `tutorials.tutorialLastSeen`, which aren't used.
- move `tutorials.setSeenTime` to `tutorial.setSeenTime`, so that it can be used with more than one tutorial in the store.
- call `tutorial.setSeenTime()` when the Get Started button is clicked.
- rewrite `updateUPP` to apply changes to the stored UPP.
- refactor `tutorials.hasNotSeenTutorialBefore` as `tutorial.hasNotBeenSeen` and add tests.

Package:
lib-classifier

Closes #2894.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

eatyourgreens commented on packages/lib-classifier/src/store/TutorialStore/Tutorial/Tutorial.js:
```diff
@@ -12,4 +12,27 @@ const Tutorial = types
+  .actions(self => {
+    return {
+      setSeenTime() {
+      const uppStore = getRoot(self).userProjectPreferences
```
Agh! broken indentation.

goplayoutside3 reviewed (changes requested):
There are new bugs with the tutorial modal in this branch. I created a brand new user account on staging and experienced the following:

While signed in with the new account, I visited the test [Transformers](https://local.zooniverse.org:3000/projects/darkeshard/test-project-2022) project for the first time.
- In the Ultraman workflow, I clicked outside of the tutorial modal to dismiss it, and the tutorial modal reappears on page refresh
- I clicked on the "Get Started" button and the tutorial modal flashes and does not close
- I clicked on the close button, and the tutorial modal disappears as expected
- Then I clicked on "Sign Out" in the nav bar, clicked on "Sign In", and the tutorial modal appears covering AuthModal

eatyourgreens commented:
I think I can reproduce that last bug on the live site. I just got this on PH TESS, after signing out and signing in.
<img width="1031" alt="The sign in for Planet Hunters TESS, but the tutorial has popped up over the log-in box, preventing me from logging in." src="https://user-images.githubusercontent.com/59547/156068759-b4cb5b9b-d5ba-4f9b-9429-f5cc66e40be6.png">

>In the Ultraman workflow, I clicked outside of the tutorial modal to dismiss it, and the tutorial modal reappears on page refresh.

This is intentional. Tutorials aren't marked as completed until you've completed them and pressed Get Started. I'll double check, but I'm fairly sure this is how the tutorial works in PFE.

eatyourgreens commented:
@goplayoutside3 I've opened #2909 for the tutorial opening when I try to sign in.

eatyourgreens commented:
Moving `tutorials.hasNotSeenTutorialBefore` to something like `tutorial.hasNotBeenSeen` might help with checking for the visibility of the popup tutorial. I'm not sure. It would help ensure that we're checking visibility for the correct tutorial ID.

eatyourgreens commented:
>I clicked on the "Get Started" button and the tutorial modal flashes and does not close.

This looks like new behaviour, presumably introduced by calling `setSeenTime()` when you click 'GetStarted'.
https://github.com/zooniverse/front-end-monorepo/blob/75094f64d65364bcb9d197c0a2867f85afa061f8/packages/lib-classifier/src/components/Classifier/components/SlideTutorial/SlideTutorial.js#L49-L52

#2893 added a check for `userProjectPreferences.loadingState` before showing the tutorial. Setting the tutorial seen time saves your preferences with the new time, which changes the UPP store's loading state, so the bug is probably there.

eatyourgreens commented:
When I click 'Get Started', while logged in, I see `hasNotSeenTutorialBefore` switch to `true`, which then prompts the tutorial to show again. There's a coupling between the UPP store state and the Tutorial store state, which isn't covered by the tests.

<img width="485" alt="Popup active state and tutorial hasNotBeenSeenBefore state after clicking 'Get Started'. hasNotBeenSeenBefore switches to true then back to false again." src="https://user-images.githubusercontent.com/59547/156141626-88d8c744-2074-49de-a6ba-f47913ee7b69.png">

eatyourgreens commented:
@goplayoutside3 we didn't add any tests for `hasNotSeenTutorialBefore` in #2893. I've expanded this PR to include the tutorial's seen state, refactored as `tutorial.hasNotBeenSeen`, and added tests.

eatyourgreens commented:
@goplayoutside3 saving the tutorial seen time changes the UPP loading state to 'putting', which then triggers the `uppStore?.loadingState !== asyncStates.success` test that was added in #2893. That then re-opens the tutorial when you mark it as seen.

That should be fixed now. I'm not seeing it locally, on a workflow with a new tutorial that I haven't seen before.

goplayoutside3 commented:
I'm still see some unexpected behavior while navigating around FEM. New user actions tested on the [Transformers](https://local.zooniverse.org:3000/projects/darkeshard/test-project-2022) project again:

While signed out:
- I’m able to navigate to the project classify page via the WorkflowSelector, make a classification, navigate to the project’s home page, then back to the classify page via the WorkflowSelector ✔️ 
- When I repeat the same actions as above, but navigate to the classify page via the nav bar, the tutorial modal for my most recently visited workflow renders on top of the WorkflowSelector modal 🚨 

While signed into an account:
- Clicking ‘Get Started’ dismisses the tutorial modal and when I navigate away from and back to the classify page, the tutorial does not reappear ✔️ 
- When I click the close button, the tutorial modal will reappear when I return to the classify page ❓ ( I think this is expected? )
- If I navigate to the classify page via the nav bar, the tutorial modal I have not dismissed with ‘Get Started’ button will render on top of the WorkflowSelector modal 🚨

eatyourgreens commented:
The tutorial appearing over the workflow selector sounds like #2909. The login form and workflow selector are part of the project app, so those bugs could be present in production.

eatyourgreens commented:
>If I navigate to the classify page via the nav bar, the tutorial modal I have not dismissed with ‘Get Started’ button will render on top of the WorkflowSelector modal 🚨

I can reproduce this on live projects. I've added a screenshot to #2909, but it's not a new bug on this branch.
https://github.com/zooniverse/front-end-monorepo/issues/2909#issuecomment-1056015200

eatyourgreens commented:
Should we record the tutorial as completed when you click outside it to dismiss it, without having gone through the steps?

eatyourgreens commented:
@goplayoutside3 I've added some thoughts to #2909 on how the NextJS app might try to stop the tutorial from appearing when there's already a modal popup open.

goplayoutside3 commented:
> Should we record the tutorial as completed when you click outside it to dismiss it, without having gone through the steps?

I think it makes sense to record that someone's completed a tutorial without stepping through. But that's just based on my use of the internet where I expect modals to go away completely when I click outside of them or use a "X" close button. This is probably a better question for Sam or Sean.

Given that the tutorial modal appearing on top of another modal is a known classifier bug, would you like review of this PR now, or wait until there's a solution for the double modal issue?

eatyourgreens commented:
>Given that the tutorial modal appearing on top of another modal is a known classifier bug, would you like review of this PR now, , or wait until there's a solution for the double modal issue?

Yes please, given that tutorial seen time isn't being recorded on live projects (https://github.com/zooniverse/front-end-monorepo/issues/2894#issuecomment-1051118251) and that's affecting new volunteers on live projects. Given that the other bug has been open for over a year, without being fixed, I wouldn't wait on that.

eatyourgreens commented:
I'll ask on the Slack project building channel about when we should consider a tutorial as completed by a volunteer. It's a one-line code change but something that project owners might have strong opinions about.

Here's a couple of related, older issues from PFE:
- Tutorials shouldn't be dismissed when you click outside them: https://github.com/zooniverse/Panoptes-Front-End/issues/1967
- Tutorial seen state should be reset after a project owner makes changes to the tutorial: https://github.com/zooniverse/Panoptes-Front-End/issues/2022

eatyourgreens commented:
I've opened #2918 to update the project app, and hopefully fix #2909.

eatyourgreens commented:
I've renamed this PR too, because `refactor tutorial setSeenTime` is really vague about what the PR actually does.

goplayoutside3 reviewed (approved):
In the scope of bug fixing UPP to save new users' tutorial seen time, this PR is approved.

I skimmed #2909 about double modals, but I also noticed the following UX while testing this branch - I'm viewing a workflow (such as [Ultraman](https://local.zooniverse.org:3000/projects/darkeshard/test-project-2022/classify/workflow/3581)) while signed out, then I sign-in with a user account I know has dismissed that workflow's tutorial, then once the AuthModal closes upon successful sign-in, the tutorial modal appears. I can replicate the same behavior with TESS.

That leads me to believe that focusing solely on preventing double modals via the NextJS app won't solve whatever is causing tutorial modals to appear on workflows where a signed-in user account should tell the modal to stay hidden. Is there an issue open where I can add documentation of this UX?

eatyourgreens commented:
That’s weird. A bug in the classify page, when it resets the classifier after you log in but the classifier store is still set up for the previous user? Does the loading state check, from #2893, get reset when the user changes, or is the user loading state still set to ‘success’? That might be a place to look for problems. 

I did a search through open issues for ‘tutorial’ but couldn’t find anything that looked like an existing issue.

I don’t think we’ve ever tested the classifier and tutorial when the user changes. My reading of the UPP code is that it expects the store to be empty when you log in.
